### ISSUE TITLE
Memory leak in 1.10.6

### ISSUE THREAD

**Issue by ajoneil** [rmosolgo/graphql-ruby#2878]
**Describe the bug**

Upgrading from 1.10.5 to 1.10.6 causes memory to climb over time - our application usually sits around 4GB of memory, it climbs to our 14GB limit in the space of about 45 minutes with 1.10.6. Reverting back to 1.10.5 resolves the issue

**Versions**

`graphql` version: 1.10.6
`graphql-pro`: 1.13.3
`rails`: 5.2.4.2
`ruby`: 2.6.6

**Comment by rmosolgo** [comment]
Yikes! Thanks for the report, I'll take a look.

**Comment by rmosolgo** [comment]
I took a quick look using the gem's `rake bench:profile_large_result` and with a test script, but I couldn't find any evidence of retained memory during query execution. 

@ajoneil , could you please help me debug this by profiling your application's schema? Here's what that entails:

- Update your app to `gem "graphql", "1.10.6"` and `bundle update graphql`
- Add `gem "memory_profiler"` to your Gemfile and `bundle install` (https://github.com/SamSaffron/memory_profiler/)
- Run a script to observe memory allocations:
  - Make `graphql_memory_profile.rb`:
  
    ```ruby 
    # graphql_memory_profile.rb 
    # Load the Rails App, assuming it's Rails: 
    require "config/environment" 

    def run_graphql_query 
      # vv Your schema and example query here 
      MyAppSchema.execute("....")
    end 

    # Warm up -- initialize some schema-level caches 
    run_graphql_query 

    # Run the profiler and print the result 
    MemoryProfiler.report { run_graphql_query }.pretty_print
    ```

   (If it's not Rails, write a script like that in some way that suits your app)

  - Execute the script, piping its output to a file: 
    
    ```
    bundle exec ruby graphql_memory_profile.rb > graphql_memory_profile.txt
    ```

- Share the result (`graphql_memory_profile.txt`) (either here, or email me directly -- redact class names, source references, etc as you see fit) 


That info would be _really_ helpful. Particularly the "retained objects" section. Locally, my profiler doesn't find any: 

```
retained objects by file
-----------------------------------
NO DATA

retained objects by location
-----------------------------------
NO DATA

retained objects by class
-----------------------------------
NO DATA
```

But of course, GraphQL schemas can be wildly different, so I probably just don't have a good reproduction!

**Comment by rmosolgo** [comment]
Thanks for sending that profile. I can see from the memory usage that you're using the legacy runtime. Here's the leak! 

https://github.com/rmosolgo/graphql-ruby/compare/v1.10.5...v1.10.6#diff-295a5500dcc6036944278a4283e3cc76R17

That's a _schema_-level cache which now uses _query contexts_ as cache keys. That means that every query the schema runs will be retained for the lifetime of the schema 😖 Oops!

I was able to replicate the issue with this script: 

```ruby 
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'graphql'
  gem 'memory_profiler'
end

module Node
  include GraphQL::Schema::Interface

  field :id, ID, null: false
end

class Thing < GraphQL::Schema::Object
  implements Node
end

class Query < GraphQL::Schema::Object
  field :node, Node, null: true do
    argument :id, ID, required: true
  end

  def node(id:)
    { id: id }
  end
end


class Schema < GraphQL::Schema
  query(Query)

  def self.resolve_type(*)
    Thing
  end
end


def run_query
  Schema.execute <<~GRAPHQL
    {
      node(id: "abcd") {
        ...on Thing {
          id
        }
      }
    }
  GRAPHQL
end

pp run_query

report = MemoryProfiler.report {
  run_query
}

report.pretty_print
```

On 1.10.6, it shows retained memory: 

```
retained objects by class
-----------------------------------
        19  Array
        17  Hash
        13  Proc
         4  String
         3  GraphQL::Filter
         2  GraphQL::Language::Nodes::Field
         1  GraphQL::Filter::MergedOnly
         1  GraphQL::Language::Nodes::Argument
         1  GraphQL::Language::Nodes::Document
         1  GraphQL::Language::Nodes::InlineFragment
         1  GraphQL::Language::Nodes::OperationDefinition
         1  GraphQL::Language::Nodes::TypeName
         1  GraphQL::Query
         1  GraphQL::Query::Context
         1  GraphQL::Query::Result
         1  GraphQL::Query::ValidationPipeline
         1  GraphQL::Schema::Warden
         1  GraphQL::StaticValidation::FieldsAreDefinedOnTypeError
         1  Method
```

But with some changes, it will go away: 

```
retained objects by class
-----------------------------------
NO DATA
```

Fix on the way at https://github.com/rmosolgo/graphql-ruby/pull/2884

**Comment by rmosolgo** [comment]
Shipped 1.10.7 -- please let me know if you run into any more trouble!
